src/io/leopard/toLeopard.ts

```diff
@@ -253,7 +422,41 @@ export default function toLeopard(
   function blockToJSWithContext(block: Block, target: Target, script?: Script): string {
     return blockToJS(block);
 
-    function inputToJS(input: BlockInput.Any): string {
+    function increase(leftSide: string, input: BlockInput.Any, allowIncrementDecrement: Boolean) {
+      let n;
+      if (input.type === "block" || ((n = Number(input.value)), isNaN(n))) {
+        return `${leftSide} += (${inputToJS(input, "number")});`;
+      }
+
+      if (allowIncrementDecrement && n === 1) {
+        return `${leftSide}++;`;
+      } else if (allowIncrementDecrement && n === -1) {
+        return `${leftSide}--;`;
+      } else if (n >= 0) {
+        return `${leftSide} += ${JSON.stringify(n)};`;
+      } else if (n < 0) {
+        return `${leftSide} -= ${JSON.stringify(-n)};`;
+      }
+    }
+
+    function decrease(leftSide: string, input: BlockInput.Any, allowIncrementDecrement: Boolean = true) {
```

Comment:
A few concerns here:
- How come `decrease` defaults `allowIncrementDecrement` to `true` but `increase` doesn't? When would we want to disallow increment/decrement operators?
- There's a lot of shared logic between `increase` and `decrease`; is it possible to combine them?
- In TypeScript, the capital-B `Boolean` [refers to a boxed `Boolean` object, which is probably not what we want for the function signature](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#number-string-boolean-symbol-and-object).

Reply:
> How come decrease defaults allowIncrementDecrement to true but increase doesn't?

That's an error. Thanks for the catch.

> There's a lot of shared logic between increase and decrease; is it possible to combine them?

There is, but these aren't conceptually heavy functions. This is a code style opinion: if two functions have little to gain from being merged into one — and would necessitate a comparatively significant amount of boilerplate — then it is probably better to leave them as separate. 

<details><summary>Long commentary on this subject</summary>
The overlap between the two functions is:

* The actual syntax tokens, i.e. `x++`, `x--`, `x += stringify(n)`, `x -= stringify(n)`
* The overall logical structure, i.e. fallback for blocks & non-number primitives then four math checks to determine syntax
* The specifics of the first step, i.e. how it detects if the input is a block or non-number primitive

The difference between the two functions is:

* When those syntax tokens are used
* The details of the third and fourth conditions (`n >= 0`, `n < 0` vs `n > 0`, `n <= 0`)
* The syntax token returned on block or non-number primitive fallback

If we tried to merge the two, we would have to face these challenges:

* `increase` and `decrease` are fundamentally distinct modes. We would have to represent which mode is desired, which requires passing an arbitrary string or enum value, adding boilerplate within the function and when calling it.
* Within the function, we would have to either find a clean way to "interweave" the associations between conditions and syntax tokens for increase/decrease, which would likely increase logical complexity, or have entirely separate branches for increase/decrease, which would add boilerplate. Either way leaves you with a function whose two second halves are functionally independent of each other and an arbitrary, non-idiomatic "switch" to decide which mode is preferred.

It is possible to move the first half of the function (fallback detection) into its own function, which would be helpful in theory since it's shared between the two. However, we actually care about two results from this first half of the code:

1. Whether or not the input was primitive *and* could be successfully parsed as a number
2. If it could be parsed as a number, what the number was

This is *possible* to represent with a single function, but requires boilerplate (and allocating more memory for multiple return values from a single function, as a short-lived array or object, if V8 doesn't optimize that out <sup>[* I touch on this again later]</sup>).

This isn't something you brought up, but the syntax I ended up using was a little awkward but was the simplest representation I could figure out, and compact. An alternative would be using an extra variable.

```ts
let n, returnCast;
if (input.type === "block") {
  returnCast = true;
} else {
  n = Number(input.value);
  if (isNaN(n)) {
    returnCast = true;
  }
}

if (returnCast) {
  return `${leftSide} ...`;
}

if (allowIncrementDecrement && n === 1) ...
```

In a separate function (meaning it's only written once), this isn't so bad, albeit a bit clunky. However, it still runs into the issue of returning both `returnCast` and `n`. Moreover, the function would only ever be used in `increase` or `decrease` and so make you wonder if it ought to never have been given a reusable form.

An alternative is to use a helper function which covers the entire structure of the function, using an anonymous callback to handle the case where the input was parsed as a number. This solves most of the problems brought up because it allows `increase` and `decrease` to still exist while not covering more logic than they strictly need to. A helper function is explicitly intended for its limited use cases, so is easier to justify than a limited-use-case "check if X can be parsed *and* see what it was parsed as" type function.(*)

```ts
function increase(leftSide, input, allowIncrementDecrement) {
  return increaseDecreaseHelper(input,
    () => `${leftSide} += (${inputToJS(input, "number")});`,
    n => {
      if (allowIncrementDecrement && n === 1) ...
    });
  // same for decrease
}

function increaseDecreaseHelper(input, handleFallback, handleNumberPrimitive) {
  let n;
  if (/* logic */) {
    return handleFallback();
  } else {
    return handleNumberPrimitive(n);
  }
}
```

(*) The actual footprint of that function could be reduced by making it return `null` if it fails to parse. This would be a "better looking" way of having two different modes of returning ("yes, it parsed, here is the value" and "no, it didn't parse, enter a different branch of code"). However, it still necessitates boilerplate for the caller (check if parsed === null, fallback if so) and doesn't change the real conceptual complexity of the function.

But the helper function still introduces boilerplate (remember the exact form of `increaseDecreaseHelper`... sure, it's just used twice, but STILL...) and just kind of looks bad.

---

The whole theme of this comment is considering the complexity *introduced by* abstracting, and whether or not hiding some other complexity is worth the added overall complexity.

One way to justify a general-purpose function with a specific arbitrary form is by having two similar functions, `parseAsString` and `parseAsNumber`, which justify the conceptual complexity by sharing the same "language" (form).

Another is just by reusing it enough. If the form is repeated enough, it gains legitimacy because the abstract form is easier to memorize than "syncing" identical changes across each place the identical logic wasn't abstracted.

Another is by the logic it replaces being sufficiently burdensome that it's worth memorizing a form to abstract it. It doesn't have to be repeated as many times to justify its existence because it's also doing the work of cutting out something that is more awkward to repeat than the form replacing it, even though it is less abstract.

I don't think we have any instances of a `parseAsString` function being useful, but there are several where the logic that would be abstracted in `parseAsNumber` is repeated. Also, it's a little awkward to repeat `(block.inputs.NUM2 as BlockInput.Any).type !== "block" && !isNaN(Number(block.inputs.NUM2.value))` every time - moreover in that case we don't even reuse `Number(block.inputs.NUM2.value)`, so that's *less* optimized code that would freely be optimized by being changed to fit the abstract form.
</details>

**TL;DR:** It's probably not worth it to combine `increase` and `decrease` or try to fully consolidate the shared logic into a helper function, but a `parseAsNumber` function which returns null for unparseables would be useful in other instances as well, and would do most of the work a more specialized helper function would anyway.